File: implementations/java/HashTable.java
public class HashTable {
    static final int SMALL      = 100;
    static final int MEDIUM     = 10000;
    static final int LARGE      = 100000;
    static final int TABLE_SIZE = 150001;

    static long[] keys     = new long[TABLE_SIZE];
    static long[] vals     = new long[TABLE_SIZE];
    static boolean[] occupied = new boolean[TABLE_SIZE];

    static void tableClear() {
        java.util.Arrays.fill(occupied, false);
    }

    static int hash(long key) {
        return (int)(((key * 2654435761L) & 0xFFFFFFFFFFFFFFFFL) % TABLE_SIZE);
    }

    static void insert(long key, long val) {
        int i = hash(key);
        while (occupied[i]) {
            if (keys[i] == key) { vals[i] = val; return; }
            i = (i + 1) % TABLE_SIZE;
        }
        keys[i] = key;
        vals[i] = val;
        occupied[i] = true;
    }

    static long lookup(long key) {
        int i = hash(key);
        while (occupied[i]) {
            if (keys[i] == key) return vals[i];
            i = (i + 1) % TABLE_SIZE;
        }
        return -1;
    }

    public static void main(String[] args) {
        int n;
        switch (args[0]) {
            case "small":  n = SMALL;  break;
            case "medium": n = MEDIUM; break;
            case "large":  n = LARGE;  break;
            default: System.err.println("Unknown size: " + args[0]); return;
        }
        tableClear();
        for (int i = 0; i < n; i++) insert((long)i * 7 + 3, (long)i * 13 + 5);
        long checksum = 0;
        for (int i = 0; i < n; i++) checksum += lookup((long)i * 7 + 3);
        System.out.println(checksum);
    }
}

File: implementations/c/hash_table.c
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <inttypes.h>

#define SMALL       100
#define MEDIUM      10000
#define LARGE       100000
#define TABLE_SIZE  150001

static int64_t keys[TABLE_SIZE];
static int64_t vals[TABLE_SIZE];
static bool    occupied[TABLE_SIZE];

static void table_clear(void) {
    memset(occupied, false, sizeof(occupied));
}

static int hash(int64_t key) {
    return (int)(((uint64_t)(key * 2654435761ULL)) % TABLE_SIZE);
}

static void insert(int64_t key, int64_t val) {
    int i = hash(key);
    while (occupied[i]) {
        if (keys[i] == key) { vals[i] = val; return; }
        i = (i + 1) % TABLE_SIZE;
    }
    keys[i]     = key;
    vals[i]     = val;
    occupied[i] = true;
}

static int64_t lookup(int64_t key) {
    int i = hash(key);
    while (occupied[i]) {
        if (keys[i] == key) return vals[i];
        i = (i + 1) % TABLE_SIZE;
    }
    return -1;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: hash_table <small|medium|large>\n");
        return 1;
    }

    int n;
    if      (strcmp(argv[1], "small")  == 0) n = SMALL;
    else if (strcmp(argv[1], "medium") == 0) n = MEDIUM;
    else if (strcmp(argv[1], "large")  == 0) n = LARGE;
    else {
        fprintf(stderr, "Unknown size: %s\n", argv[1]);
        return 1;
    }

    table_clear();
    for (int i = 0; i < n; i++) {
        insert((int64_t)i * 7 + 3, (int64_t)i * 13 + 5);
    }

    int64_t checksum = 0;
    for (int i = 0; i < n; i++) {
        checksum += lookup((int64_t)i * 7 + 3);
    }

    printf("%" PRId64 "\n", checksum);
    return 0;
}
